### PR TITLE
Added interfaces in preparation to refactoring NavigationService

### PR DESCRIPTION
**Pull request by olegzil**
## Description
In preparation to refactor NavigationService I've added a set of new interfaces. 
TripService is a reimplementation of NavigationService, but without reliance on other classes
The problem is that the current implementation of NavigationService needs an instance of MapboxNavigator, a class that will be removed. It uses that instance internally and also, passes it as a parameter to other classes.
You can’t just pull that out. Everything is going to break. So, the first step is to create an interface for MapboxNavigation. This interface includes only public classes. Once you have that interface, you can cut the connection between the NavigationService, NavigationEventDispatcher and MapboxNavigator. You can also use this interface in other classes that NavigationService creates. Next, you have to create interfaces for  MapboxNavigator and RouteRefresher. The reason is that the two classes are marked internal and MapboxNavigator has functions that return references to these classes. Since the interface is public, you cannot do this. By adding the two interfaces I just described, you avoid that problem. Finally, we’ll have to break up the MapboxNavigation interface into smaller once, along areas of responsibility so that you don’t have one giant interface. The MapboxNavigator interface currently has 53 methods. That’s not good. So we’ll break that up into functional groups. My next PR is to going to be these interfaces without much code change otherwise, then once I break that up I’ll finally update NavigationService with the desired behavior.

### Goal
Preview 

Please describe the PR goals. Just the stuff needed to implement the fix / feature and a simple rationale. It could contain many check points if needed

### Implementation

Please include all the relevant things implemented and also rationale, clarifications / disclaimers etc. related to the approach used. It could be as self code companion comments

## Screenshots or Gifs

Please include all the media files to give some context about what's being implemented or fixed. It's not mandatory to upload screenshots or gifs, but for most of the cases it becomes really handy to get into the scope of the feature / bug being fixed and also it's REALLY useful for UI related PRs screenshot gif

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/MapboxNavigation.java [hunk 1]
```diff
@@ -68,7 +68,7 @@
  * @see <a href="https://www.mapbox.com/android-docs/navigation/">Navigation documentation</a>
  * @since 0.1.0
  */
-public class MapboxNavigation implements ServiceConnection {
+public class MapboxNavigation implements ServiceConnection, MapboxNavigationInterface {
```
Comment (kmadsen) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/MapboxNavigation.java]: Opinion: give interfaces the clean name and the implementation the extra. We'll maybe hide the implementation of the interface.

interface MapboxNavigation
class MapboxNavigationImpl

interface LocationProvider
class AndroidLocationProvider, MockLocationProvider

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationEventDispatcherInterface.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import com.mapbox.services.android.navigation.v5.milestone.Milestone
+import com.mapbox.services.android.navigation.v5.milestone.MilestoneEventListener
+import com.mapbox.services.android.navigation.v5.navigation.EnhancedLocationListener
+import com.mapbox.services.android.navigation.v5.navigation.NavigationEventListener
+import com.mapbox.services.android.navigation.v5.offroute.OffRouteListener
+import com.mapbox.services.android.navigation.v5.route.FasterRouteListener
+import com.mapbox.services.android.navigation.v5.routeprogress.ProgressChangeListener
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
+
+interface NavigationEventDispatcherInterface {
```
Comment (RingerJK) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationEventDispatcherInterface.kt]: @olegzil it's a really huge interface, would you like to split it to smallest? it stays more readable, of course up 2 u.

Comment (olegzil) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationEventDispatcherInterface.kt]: If you look at the comment "// Interface boundary" you'll see the proposal as to where this interface can be partitioned.

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExampleViewModel.kt [hunk 3]
```diff
@@ -65,7 +65,7 @@ class ExampleViewModel(application: Application) : AndroidViewModel(application)
 
         // Initialize navigation and pass the LocationEngine
         navigation = MapboxNavigation(getApplication(), accessToken)
-        navigation.locationEngine = locationEngine
```
Comment (kmadsen) [inline app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExampleViewModel.kt]: kotlin interfaces could continue to support `val`

wonder what makes us decide one way or another here

**Comment by kmadsen** [comment]
I'm on board with using interfaces to help us decompose classes with many responsibilities. Also would make it easier to add unit tests. 👍

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExampleViewModel.kt (hunk 3 above)
Comment (olegzil) [reply app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExampleViewModel.kt]: Good question. I haven't used interfaces that also contain val. So I default to using pure interfaces. But if a case can be made to add val to an interface, I have no issues with that.

**Pull request closed without merging**